### PR TITLE
SNS should resolve domains like b.sol

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/35405

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/json_rpc_service_unittest.cc [hunk 1]
```diff
@@ -3657,19 +3657,54 @@ TEST_F(JsonRpcServiceUnitTest, GetWalletAddrInvalidDomain) {
...
+  std::vector<std::string> valid_domains = {
+      "brave.sol",                //
+      "test.brave.sol",           //
+      "brave-test.test-dev.sol",  //
+      "b.sol",                    //
+      "w.sol",                    //
+      "-.sol",                    //
+      "-brave.sol",               //
+      "brave-.sol",               //
+      "---.sol",                  //
```
Comment (supermassive) [inline components/brave_wallet/browser/json_rpc_service_unittest.cc:3687]: @kdenhartog 
Is it ok to support resolving of such hostnames like `-.sol`?
Such names are valid from their side https://www.sns.id/domain?domain=-----

**Review by yrliou — APPROVED** [review]

File: components/brave_wallet/browser/json_rpc_service_unittest.cc (hunk 1 above)
Comment (kdenhartog) [reply components/brave_wallet/browser/json_rpc_service_unittest.cc:3687]: I'll need to double check the URL spec at WHATWG. Do all of these correctly validate with `GURL::is_valid()`? That would be my primary concern and then making sure the domains match up with URL spec and RFC3986 would be my secondary concerns.

**Comment by kdenhartog** [comment]
I think this should probably be fine, but I'll need to read over specs to check some edge cases.

File: components/brave_wallet/browser/json_rpc_service_unittest.cc (hunk 1 above)
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service_unittest.cc:3687]: Well https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2 says 
```
Such a name consists of a sequence of domain labels separated by ".", each domain label starting and ending with an alphanumeric character and possibly also containing "-" characters.
```
This is ok
```
EXPECT_TRUE(GURL("https://-.sol").is_valid());
```
This also works
image

File: components/brave_wallet/browser/json_rpc_service.cc [hunk 2]
```diff
@@ -1994,8 +1997,14 @@ void JsonRpcService::OnGetBlockByNumber(GetBlockByNumberCallback callback,
...
-bool JsonRpcService::IsValidDomain(const std::string& domain) {
-  static const base::NoDestructor<re2::RE2> kDomainRegex(kDomainPattern);
+bool JsonRpcService::IsValidEnsDomain(const std::string& domain) {
+  static const base::NoDestructor<re2::RE2> kDomainRegex(kEnsDomainPattern);
+  return re2::RE2::FullMatch(domain, *kDomainRegex);
+}
+
+/*static*/
+bool JsonRpcService::IsValidSnsDomain(const std::string& domain) {
+  static const base::NoDestructor<re2::RE2> kDomainRegex(kSnsDomainPattern);
```
Comment (kdenhartog) [inline components/brave_wallet/browser/json_rpc_service.cc:2007]: Should we lowercase this domain before checking the kDomainRegex? I'm not finding documentation of SNS supported characterset, but this seems a bit overly restrictive.

Comment (kdenhartog) [reply components/brave_wallet/browser/json_rpc_service.cc:2007]: Screenshot 2024-01-22 at 4 29 46 PM

File: components/brave_wallet/browser/json_rpc_service_unittest.cc (hunk 1 above)
Comment (kdenhartog) [reply components/brave_wallet/browser/json_rpc_service_unittest.cc:3687]: Ok, thanks for confirming. After looking through this further this makes sense. There's some similar cases when dealing with punycode conversions from emojis and looking at the definition, the host is non-empty and it's using valid ASCII code points so this checks out to me.

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 11 x64` using the following build(s):

```
Brave | 1.64.17 Chromium: 121.0.6167.85 (Official Build) nightly (64-bit)
-- | --
Revision | 20b7d9bbdbe80199a28d59b921ba84f92c787325
OS | Windows 11 Version 23H2 (Build 22631.3007)
```

Using the STR/Cases outlined via https://github.com/brave/brave-browser/issues/35405#issue-2088020324, reproduced the original issue as per the following:

reproducedSol

Using the STR/Cases outlined via https://github.com/brave/brave-browser/issues/35405#issue-2088020324, ensured that `w.sol` is being resolved correctly as per the following:

https://github.com/brave/brave-core/assets/2602313/46e234fc-33ec-44dc-a368-e78b8e0b2366

**Pull request merged**